Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php

```diff
@@ -312,6 +314,68 @@ public function migrateMetaDataToWorkspaceName(\Closure $outputFn)
         $outputFn(sprintf('Migration applied to %s events.', count($this->eventsModified)));
     }
 
+    /**
+     *  Adds the "workspaceName" to the data of all content stream related events
+     *
+     *  Needed for feature "Add workspaceName to relevant events": https://github.com/neos/neos-development-collection/issues/4996
+     *
+     *  Included in May 2024 - before final Neos 9.0 release
+     *
+     * @param \Closure $outputFn
+     * @return void
+     */
+    public function migratePayloadToWorkspaceName(\Closure $outputFn): void
+    {
+
+        $backupEventTableName = DoctrineEventStoreFactory::databaseTableName($this->contentRepositoryId) . '_bkp_' . date('Y_m_d_H_i_s');
+        $outputFn('Backup: copying events table to %s', [$backupEventTableName]);
+        $this->copyEventTable($backupEventTableName);
+
+        $numberOfMigratedEvents = 0;
+        $workspaceNamesByContentStreamId = [];
+
+        foreach ($this->eventStore->load(VirtualStreamName::all()) as $eventEnvelope) {
+            $eventType = $eventEnvelope->event->type->value;
+            if (in_array($eventType, ['RootWorkspaceWasCreated', 'WorkspaceBaseWorkspaceWasChanged', 'WorkspaceWasCreated', 'WorkspaceWasDiscarded', 'WorkspaceWasPartiallyDiscarded', 'WorkspaceWasRebased'], true)) {
+                $eventData = self::decodeEventPayload($eventEnvelope);
+                $workspaceNamesByContentStreamId[$eventData['newContentStreamId']] = $eventData['workspaceName'];
+                continue;
+            }
+            if (in_array($eventType, ['WorkspaceWasPartiallyPublished', 'WorkspaceWasPublished'], true)) {
+                $eventData = self::decodeEventPayload($eventEnvelope);
+                $workspaceNamesByContentStreamId[$eventData['newSourceContentStreamId']] = $eventData['sourceWorkspaceName'];
+                continue;
+            }
+            if (!in_array($eventType, ['DimensionShineThroughWasAdded', 'DimensionSpacePointWasMoved', 'NodeAggregateWithNodeWasCreated', 'NodeAggregateWasDisabled', 'NodeAggregateWasEnabled', 'NodePropertiesWereSet', 'NodeAggregateWasMoved', 'NodeReferencesWereSet', 'NodeAggregateWasRemoved', 'NodeAggregateNameWasChanged', 'NodeAggregateTypeWasChanged', 'NodeGeneralizationVariantWasCreated', 'NodePeerVariantWasCreated', 'NodeSpecializationVariantWasCreated', 'RootNodeAggregateDimensionsWereUpdated', 'RootNodeAggregateWithNodeWasCreated', 'SubtreeWasTagged', 'SubtreeWasUntagged'], true)) {
+                continue;
+            }
+            $eventData = self::decodeEventPayload($eventEnvelope);
+            $this->updateEventPayload($eventEnvelope->sequenceNumber, [...$eventData, 'workspaceName' => $workspaceNamesByContentStreamId[$eventData['contentStreamId']]]);
```

Comment:
> Warning: Undefined array key "9cc426e4-3520-4e48-8272-609267850177" in EventMigrationService.php line 353

this time luck is at our side. I happened to have the events from this one guy still in the db and accidetally applied the migration to it ^^

lucky for us, because it seems that if there is no workspace for `contentStreamId` this line will fail.

In my case it seems to be a simple content stream that i suppose was created just to set properties and remove. This seems to be the footprint of either structure adjustment or node migrations.
However we need to properly handle this and write `missing` as ws name into it.

Reply:
good call, what do you think of my pragmatic fix? :)

Reply:
exactly as i intended it ^^